File: microstream/firmware/src/Protocol.cpp
#include "Protocol.h"

namespace MicrostreamProtocol {

  unsigned int encode(uint8_t* buffer, MessageType type, const uint8_t* payload, unsigned int payloadLen) {
    buffer[0] = (uint8_t)type;
    if (payload && payloadLen > 0) {
      memcpy(buffer + 1, payload, payloadLen);
    }
    return 1 + payloadLen;
  }

  unsigned int encodeAudioData(uint8_t* buffer, const uint8_t* samples, unsigned int sampleCount) {
    return encode(buffer, AUDIO_DATA, samples, sampleCount);
  }

  unsigned int encodeAudioEnd(uint8_t* buffer) {
    return encode(buffer, AUDIO_END, NULL, 0);
  }

  unsigned int encodeHeartbeat(uint8_t* buffer) {
    return encode(buffer, HEARTBEAT, NULL, 0);
  }

  MessageType decodeType(const uint8_t* data) {
    return (MessageType)data[0];
  }

  const uint8_t* decodePayload(const uint8_t* data) {
    return data + 1;
  }
}

File: microstream/server/src/Protocol.js
/**
 * Binary message protocol for microstream.
 *
 * Each WebSocket binary message:
 *   [ type (1 byte) | payload (0-N bytes) ]
 *
 * WebSocket handles framing, so no length field is needed.
 */

const MessageType = {
  AUDIO_DATA: 0x01,
  AUDIO_END: 0x02,
  HEARTBEAT: 0x03,
  CONFIG: 0x04,
  ERROR: 0x05
}

function encode (type, payload) {
  const payloadBuf = payload ? Buffer.from(payload) : Buffer.alloc(0)
  const msg = Buffer.allocUnsafe(1 + payloadBuf.length)
  msg.writeUInt8(type, 0)
  if (payloadBuf.length > 0) {
    payloadBuf.copy(msg, 1)
  }
  return msg
}

function decode (data) {
  const buf = Buffer.from(data)
  if (buf.length < 1) {
    return null
  }
  return {
    type: buf.readUInt8(0),
    payload: buf.length > 1 ? buf.slice(1) : Buffer.alloc(0)
  }
}

function encodeAudioData (pcmChunk) {
  return encode(MessageType.AUDIO_DATA, pcmChunk)
}

function encodeAudioEnd () {
  return encode(MessageType.AUDIO_END)
}

function encodeHeartbeat () {
  return encode(MessageType.HEARTBEAT)
}

function encodeConfig (audioConfig) {
  const json = JSON.stringify(audioConfig)
  return encode(MessageType.CONFIG, Buffer.from(json, 'utf8'))
}

function encodeError (reason) {
  return encode(MessageType.ERROR, Buffer.from(reason, 'utf8'))
}

module.exports = {
  MessageType,
  encode,
  decode,
  encodeAudioData,
  encodeAudioEnd,
  encodeHeartbeat,
  encodeConfig,
  encodeError
}
